Make RM_Yield thread-safe

sundb opened the pull request:
## Issues and solutions from #12817
1. Touch ProcessingEventsWhileBlocked and calling moduleCount() without GIL in afterSleep()
    - Introduced: 
       Version: 7.0.0
       PR: #9963

   - Harm Level: Very High
     If the module thread calls `RM_Yield()` before the main thread enters afterSleep(), 
     and modifies `ProcessingEventsWhileBlocked`(+1), it will cause the main thread to not wait for GIL,
     which can lead to all kinds of unforeseen problems, including memory data corruption.

   - Initial / Abandoned Solution:
      * Added `__thread` specifier for ProcessingEventsWhileBlocked.
   	     `ProcessingEventsWhileBlocked` is used to protect against nested event processing, but event processing
         in the main thread and module threads should be completely independent and unaffected, so it is safer
         to use TLS.
      * Adding a cached module count to keep track of the current number of modules, to avoid having to use `dictSize()`.
    
    - Related Warnings:
```
WARNING: ThreadSanitizer: data race (pid=1136)
  Write of size 4 at 0x0001045990c0 by thread T4 (mutexes: write M0):
    #0 processEventsWhileBlocked networking.c:4135 (redis-server:arm64+0x10006d124)
    #1 RM_Yield module.c:2410 (redis-server:arm64+0x10018b66c)
    #2 bg_call_worker <null>:83232836 (blockedclient.so:arm64+0x16a8)

  Previous read of size 4 at 0x0001045990c0 by main thread:
    #0 afterSleep server.c:1861 (redis-server:arm64+0x100024f98)
    #1 aeProcessEvents ae.c:408 (redis-server:arm64+0x10000fd64)
    #2 aeMain ae.c:496 (redis-server:arm64+0x100010f0c)
    #3 main server.c:7220 (redis-server:arm64+0x10003f38c)
```

2. aeApiPoll() is not thread-safe
    When using RM_Yield to handle events in a module thread, if the main thread has not yet
    entered `afterSleep()`, both the module thread and the main thread may touch `server.el` at the same time.

    - Introduced: 
       Version: 7.0.0
       PR: #9963

   - Old / Abandoned Solution:
      Adding a new mutex to protect timing between after beforeSleep() and before afterSleep().
      Defect: If the main thread enters the ae loop without any IO events, it will wait until
                   the next timeout or until there is any event again, and the module thread will
                   always hang until the main thread leaves the event loop.

    - Related Warnings:
```
SUMMARY: ThreadSanitizer: data race ae_kqueue.c:55 in addEventMask
==================
==================
WARNING: ThreadSanitizer: data race (pid=14682)
  Write of size 4 at 0x000100b54000 by thread T9 (mutexes: write M0):
    #0 aeApiPoll ae_kqueue.c:175 (redis-server:arm64+0x100010588)
    #1 aeProcessEvents ae.c:399 (redis-server:arm64+0x10000fb84)
    #2 processEventsWhileBlocked networking.c:4138 (redis-server:arm64+0x10006d3c4)
    #3 RM_Yield module.c:2410 (redis-server:arm64+0x10018b66c)
    #4 bg_call_worker <null>:16042052 (blockedclient.so:arm64+0x169c)

  Previous write of size 4 at 0x000100b54000 by main thread:
    #0 aeApiPoll ae_kqueue.c:175 (redis-server:arm64+0x100010588)
    #1 aeProcessEvents ae.c:399 (redis-server:arm64+0x10000fb84)
    #2 aeMain ae.c:496 (redis-server:arm64+0x100010da8)
    #3 main server.c:7238 (redis-server:arm64+0x10003f51c)
```

## The final fix as the comments: https://github.com/redis/redis/pull/12817#discussion_r1436427232
Optimized solution based on the above comment:

First, we add `module_gil_acquring` to indicate whether the main thread is currently in the acquiring GIL state.

When the module thread starts to yield, there are two possibilities(we assume the caller keeps the GIL):
1. The main thread is in the mid of beforeSleep() and afterSleep(), that is, `module_gil_acquring` is not 1 now.
    At this point, the module thread will wake up the main thread through the pipe and leave the yield,
    waiting for the next yield when the main thread may already in the acquiring GIL state. 
    
2. The main thread is in the acquiring GIL state.
    The module thread release the GIL, yielding CPU to give the main thread an opportunity to start
     event processing, and then acquire the GIL again until the main thread releases it.
     This is what https://github.com/redis/redis/pull/12817#discussion_r1436427232 mentioned direction.

oranagra commented on src/module.c:
```diff
@@ -2394,6 +2394,14 @@ void RM_Yield(RedisModuleCtx *ctx, int flags, const char *busy_reply) {
+            if (!pthread_equal(server.main_thread_id, pthread_self())) {
+                int acquiring;
+                atomicGet(server.module_gil_acquring, acquiring);
+                if (!acquiring) goto end; /* We can do nothing if the main thread is not yet in acquiring GIL state. */
```
we can write to the `server.module_pipe` to wake it up, and then wait for it to get to the right state..
but still, considering the context (RM_Yield is meant to give redis a chance to process events, not to force it to do so), maybe it would better serve the caller if we won't wait (don't block the caller just yet).

but let's imagine a server with a real low hz (e.g. 1), and no client traffic.
this means the server will not wake up anytime soon. so maybe we should write to the pipe, and return from RM_Yield, so that in the next call, we're likely to already be waiting for the GIL in the other thread?

on the other hand, in the scenario i described above, if there are no clients waiting to be served, then there's no need to call processEventsWhileBlocked.

WDYT?

sundb replied:
> on the other hand, in the scenario i described above, if there are no clients waiting to be served, then there's no need to call processEventsWhileBlocked.

but even if there are no clients, we still need to accept new connections,
thus, we still need to call `processEventsWhileBlocked`.

sundb replied:
> we can write to the `server.module_pipe` to wake it up, and then wait for it to get to the right state.. but still, considering the context (RM_Yield is meant to give redis a chance to process events, not to force it to do so), maybe it would better serve the caller if we won't wait (don't block the caller just yet).

I prefer not to block the caller, if the caller can't handle the events at the moment, the caller can continue to do what it hasn't done before, and come back later.

oranagra replied:
> but even if there are no clients, we still need to accept new connections

that's ok, we'll get an event on the listener socket

oranagra reviewed (commented):
thanks.
please extend the top comment with:
1. description of the original issue (it's good to have a reference to the previous discussion, but let's have a summary of the issue described here too).
2. if we're gonna refer to a previous idea that turned out being wrong, and list why it was wrong, we should also describe that idea. or we can just avoid mentioning it altogether.
3. maybe we can list some of the considerations of the comment i added below if / when we agree on the details (mention them in both the code and the top comment)

shahsb commented on src/module.c:
```diff
@@ -2394,6 +2394,24 @@ void RM_Yield(RedisModuleCtx *ctx, int flags, const char *busy_reply) {
+            /* If it is not the main thread, we need to process events after the
+             * main thread enter acquiring GIL state in order to protect the ae and
+             * avoid potential race conditions. */
+            if (!pthread_equal(server.main_thread_id, pthread_self())) {
```
```suggestion
            if ( 0 == pthread_equal(server.main_thread_id, pthread_self()) ) {
```
Please incorporate this for better readability. 

Also, please always follow this. 

Remember, Positive always overtakes negative (being optimistic). While you read as well as when you think in real life as well.

sundb replied:
Thanks, but I think using `!` is a very simple and readable way, at least it saves 2 characters.

shahsb commented on src/module.c:
```diff
@@ -2394,6 +2394,24 @@ void RM_Yield(RedisModuleCtx *ctx, int flags, const char *busy_reply) {
+            if (!pthread_equal(server.main_thread_id, pthread_self())) {
+                int acquiring;
+                atomicGet(server.module_gil_acquring, acquiring);
+                if (!acquiring) {
```
```suggestion
                if ( acquiring == false ) {
```

sundb replied:
Same as above.
btw: we dont use `false` in Redis.

shahsb commented on src/module.c:
```diff
@@ -2394,6 +2394,24 @@ void RM_Yield(RedisModuleCtx *ctx, int flags, const char *busy_reply) {
+                    if (write(server.module_pipe[1],"A",1) != 1) {
+                        /* Ignore the error, this is best-effort. */
+                    }
+                    goto end;
```
Using goto is highly discouraged.

It leads to spaghetti code and unmaintainable mess.

Please always refactor such codes. Also, if you find existing occurrences,, kindly correct them as well.

Always strive to put things in a better state.

sundb replied:
there is nothing wrong with `goto` if we using it properly.
We can quickly jump out of the loop, or clean up before leaving method, which is much cleaner than using intermediate variables or using `do {} while (0)`, isn't it?

shahsb replied:
You don't need any intermediate variables or do-while loop here. 

At the end, you are just doing `yield_nesting--;` for that you do not need goto, intermediate variables or do-while loop.

that you can do simply here and return.   `yield_nesting--; return;`  THAT'S IT.

why to unnecessarily complicate things?

Using goto is considered poor practice and highly discouraged industry wide. 

Remember, the goal of good programming is to write code that is not only functionally correct but also readable, maintainable, and well-structured. By avoiding overreliance on goto and utilizing alternative control flow mechanisms, you can achieve these qualities and enhance the overall quality of the program.

Prioritize structured programming principles for better readability, maintainability, and code clarity.

shahsb replied:
there is extensive list and research done for this, some of them I can share for your reference:

Readability and Maintainability: Code that uses goto statements can become hard to read and understand. It can lead to "spaghetti code," where the flow of execution jumps around in a non-linear fashion, making it difficult for programmers (including yourself) to follow and maintain the code.

Debugging Complexity: Code with goto statements can make debugging more challenging. When an issue arises, tracing the flow of execution becomes more complex, making it harder to identify and fix bugs.

Structured Programming Principles: The use of goto goes against the principles of structured programming, which advocates for modular and organized code structures. Modern programming languages provide constructs like loops, functions, and conditional statements, which make it possible to achieve the same functionality without resorting to goto.

Portability: Code that relies on goto may be less portable across different platforms and compilers. Some systems may not support certain uses of goto, leading to potential issues when trying to run the code on different environments.

Code Quality and Best Practices: In most cases, there are alternative constructs (such as loops and conditionals) that can be used to achieve the same results in a clearer and more maintainable way. Emphasizing the use of these constructs is considered a best practice for writing high-quality code.

Structured Exception Handling: In languages that support structured exception handling (like try-catch blocks), using goto can interfere with these mechanisms and make error handling less straightforward.

Structure: Excessive goto usage can break down the natural structure of your program. Instead of clear, sequential flow, you might end up with spaghetti code – tangled and hard to untangle. This can further reduce readability and maintainability.

shahsb replied:
General principles that I follow:

Structured programming: Prioritize using control flow constructs like loops, conditionals, and functions to maintain clear and predictable code flow.

Readability: Aim for code that is easily understandable by you and others. Avoid confusing logic and tangled jumps that rely on goto.

Refactoring: If you encounter existing code that heavily uses goto, consider revisiting and restructuring it using the alternatives mentioned above.

Functions: Breaking down your code into smaller, well-defined functions can improve readability and modularity. If you find yourself using goto to jump around within a single function, consider extracting the targeted code into a separate function and calling it where needed.

Early return: This allows you to exit a function immediately if certain conditions are met, bypassing further code execution. It can be used as an alternative to jumping to a specific point within the function using goto.

sundb commented:
@oranagra I dropped the last solution and used the one you give in https://github.com/redis/redis/pull/12817#discussion_r1436427232

the reason why moving the events processing into main thread:

The main thread has prepared the eventloop's io events and the events number
before it enters the acquiring GIL state, and when the module thread enters the event
loop, it will rebuild this data and process these events data.

Note that at this point the data for the main thread and the module thread may be in two caches,
and we don't have a memory barrier to protect them.

When the main thread comes back, the events it was processing may have already been processed or invalid.

oranagra commented:
didn't you have some argument against that approach at some point? or was that argument dismissed once we started skipping RM_Yield until we know the main thread is already blocked on the GIL?

sundb commented:
@oranagra Haha, this last solution is actually a combination of the original solution, it avoids idling before the main thread goes in to acquire the gil, and then uses sched_yield instead of usleep to more aggressively yield the cpu slice.

sundb commented:
@oranagra Haha, this last solution is actually a combination of the original solution, it avoids idling before the main thread goes in to acquire the gil, and then uses sched_yield instead of usleep to more aggressively yield the cpu slice.

shahsb commented on src/module.c:
```diff
@@ -2407,7 +2407,34 @@ void RM_Yield(RedisModuleCtx *ctx, int flags, const char *busy_reply) {
+                    if (write(server.module_pipe[1],"A",1) != 1) {
+                        /* Ignore the error, this is best-effort. */
+                    }
+                    goto end;
```
```suggestion
                     yield_nesting--; return;
```

sundb replied:
Agree with you, `goto` can be abused, but we can't ignore its advantages because of its disadvantages.
1. simplifies error handling, easy to jump to error block.
2. jumping out of multiple nesting, although we can accomplish the same thing with other ways, but `goto` is more simple.
3. improves readability, and allows you to reuse duplicate code, as in the current example, so that if we need to exit  in another place in the future, all we need to do is `goto end`.

sundb commented:
@oranagra I fixed a bug that  missing restoring busy_module_yield_reply and busy_module_yield_flags.
I've updated the top comment and it's ready for merge now.

sundb commented:
@oranagra Please wait for me, right now the moduleapi test gets stuck, I'm not sure if it's caused by merging unstable with #12857.

sundb commented:
@oranagra this PR can be safely merged.

oranagra commented:
@sundb thanks a lot.
considering the fix is rather simple, it looks safe to backport this (at least to 7.2), maybe after letting it sit in unstable for a few weeks.
WDYT?

sundb commented:
@oranagra sure.
